File: lib/features/offers/domain/entities/offer_entity.dart
import 'package:equatable/equatable.dart';

enum DiscountType { PERCENTAGE, FIXED }

class OfferEntity extends Equatable {
  final String id;
  final String name;
  final String description;
  final DiscountType discountType;
  final double discountValue;
  final DateTime startDate;
  final DateTime endDate;
  final bool isActive;
  final String merchantId;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<OfferProductEntity> products;
  final OfferMerchantEntity merchant;

  const OfferEntity({
    required this.id,
    required this.name,
    required this.description,
    required this.discountType,
    required this.discountValue,
    required this.startDate,
    required this.endDate,
    required this.isActive,
    required this.merchantId,
    required this.createdAt,
    required this.updatedAt,
    this.products = const [],
    required this.merchant,
  });

  @override
  List<Object?> get props => [
        id,
        name,
        description,
        discountType,
        discountValue,
        startDate,
        endDate,
        isActive,
        merchantId,
        createdAt,
        updatedAt,
        products,
        merchant,
      ];
}

class OfferProductEntity extends Equatable {
  final String id;
  final String name;
  final double price;
  final double? offerPrice;

  const OfferProductEntity({
    required this.id,
    required this.name,
    required this.price,
    this.offerPrice,
  });

  @override
  List<Object?> get props => [id, name, price, offerPrice];
}

class OfferMerchantEntity extends Equatable {
  final String id;
  final String restaurantName;

  const OfferMerchantEntity({
    required this.id,
    required this.restaurantName,
  });

  @override
  List<Object?> get props => [id, restaurantName];
}

File: Jeeb_NestJs-master/src/database/entities/offer.entity.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToOne,
  OneToMany,
  ManyToMany,
  OneToMany as OneToManyAlt,
  JoinColumn,
  CreateDateColumn,
  UpdateDateColumn,
} from 'typeorm';
import { User } from './user.entity';
import { Product } from './product.entity';
import { Order } from './order.entity';
import { OfferProduct } from './offer-product.entity';
import { Image } from './image.entity';
import { DiscountType } from '../../common/enums';

@Entity('offers')
export class Offer {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({
    type: 'jsonb',
    comment: 'Stores name in { ar: string, en: string }',
    transformer: {
      to(value: string): any {
        return { ar: value };
      },
      from(value: any): string {
        return value?.ar || value;
      },
    },
  })
  name: string;

  @Column({
    type: 'jsonb',
    nullable: true,
    transformer: {
      to(value: string): any {
        return { ar: value };
      },
      from(value: any): string {
        return value?.ar || value;
      },
    },
  })
  description: string | null;

  @Column({
    type: 'enum',
    enum: DiscountType,
  })
  discountType: DiscountType;

  @Column({ type: 'float' })
  discountValue: number;

  @Column({ type: 'timestamp', nullable: true })
  startDate: Date | null;

  @Column({ type: 'timestamp', nullable: true })
  endDate: Date | null;

  @Column({ default: true })
  isActive: boolean;

  @Column({ nullable: true })
  merchantId: number | null;

  @ManyToOne(() => User, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'merchantId' })
  merchant: User;

  @OneToMany(() => OfferProduct, (op) => op.offer)
  offerProducts: OfferProduct[];

  @OneToManyAlt(() => Image, (image) => image.entityId, {
    onDelete: 'CASCADE',
  })
  images: Image[];

  @ManyToMany(() => Order, (order) => order.offers)
  orders: Order[];

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;
}
